doc/developer/design/20230705_v2_txn_management.md

```diff
@@ -0,0 +1,422 @@
+txn.commit_at(&mut txns, 4).await.expect("ts 4 available")
+    .apply(&mut txns).await;
+
+// Read data shard(s) at some `read_ts`.
+let updates = d1_read.snapshot_and_fetch(
+    vec![txns.read_cache().to_data_inclusive(&d1, 4).unwrap()].into()
+).await.unwrap();
+```
+
+### Writes
+
+The structure of the txns shard is `(ShardId, Vec<u8>)` updates.
+
+The core mechanism is that a txn commits a set of transmittable persist _batch
+handles_ as `(ShardId, <opaque blob>)` pairs at a single timestamp. This
+contractually both commits the txn and advances the logical upper of _every_
+data shard (not just the ones involved in the txn).
+
+Example:
+
+```text
+// A txn to only d0 at ts=1
+(d0, <opaque blob A>, 1, 1)
+// A txn to d0 (two blobs) and d1 (one blob) at ts=4
+(d0, <opaque blob B>, 4, 1)
+(d0, <opaque blob C>, 4, 1)
+(d1, <opaque blob D>, 4, 1)
+```
+
+However, the new commit is not yet readable until the txn apply has run, which
+is expected to be promptly done by the committer, except in the event of a
+crash. This, in ts order, moves the batch handles into the data shards with a
+`compare_and_append_batch` (similar to how the multi-worker persist_sink works).
+
+Once apply is run, this is noted in the txns shard by retracting the update
+adding the batch. As a result, the contents of the txns shard at any given
+timestamp is exactly the set of outstanding apply work.
+
+Example (building on the above):
+
+```text
+// Apply for the first txn at ts=3
+(d0, <opaque blob A>, 3, -1)
+// Apply for the second txn (the timestamps can be different for each
+// retraction in a txn, but don't need to be)
+(d0, <opaque blob B>, 5, -1)
+(d0, <opaque blob C>, 6, -1)
+(d1, <opaque blob D>, 6, -1)
+```
+
+To make it easy to reason about exactly which data shards are registered in the
+txn set at any given moment, the data shard is added to the set with a
+`(ShardId, <empty>)` pair. The data may not be read before the timestamp of the
+update (which starts at the time it was initialized, but it may later be
+forwarded).
+
+Example (building on both of the above):
+
+```text
+// d0 and d1 were both initialized before they were used above
+(d0, <empty>, 0, 1)
+(d1, <empty>, 2, 1)
+```
+
+### Reads
+
+Reads of data shards are almost as straightforward as writes. A data shard may
+be read normally, using snapshots, subscriptions, shard_source, etc, through the
+most recent non-empty write. However, the upper of the txns shard (and thus the
+logical upper of the data shard) may be arbitrarily far ahead of the physical
+upper of the data shard. As a result, we have to "translate" timestamps by
+synthesizing ranges of empty time that have not yet been committed (but are
+guaranteed to remain empty):
+
+- To take a snapshot of a data shard, the `as_of` is passed through unchanged if
+  the timestamp of that shard's latest non-empty write is past it. Otherwise,
+  the timestamp of the latest non-empty write is used.
```

Comment:
What if the `as_of` is past the latest non-empty write? For example:

- Shard A's latest non-empty write is `10`.
- The txn shard's latest non-empty write is `20`.
- Someone requests a snapshot of shard A as of `30`?

I think the correct thing to do here is to use a timestamp of `30`, not `20` which is what I think this is suggesting?

As a follow-up, if we do choose a timestamp of `30` and shard A never gets updated again, but the txn shard eventually advances past `30`, then how do we find this out and complete the read?

Reply:
I'm a bit confused why the last non-empty write at `20` is relevant, but if I'm following the question correctly, it's actually suggesting `10`.

If the last non-empty write to some data shard is at `10`, and the txn shard's upper is past some snapshot as_of `X`, then we know that `[10,X]` is empty in A, so it's safe to just do a snapshot directly on the shard at `10`. That's true even if we haven't yet finished applying everything in A: e.g. if the write at `10` hasn't applied yet then the snapshot will just block until it does.

Otoh, if the txn shard's upper is not past `X` yet, then anything could happen and we have to wait until it does before saying anything about the snapshot (which is normally what happens with snapshots).

Reply:
> I'm a bit confused why the last non-empty write at `20` is relevant, but if I'm following the question correctly, it's actually suggesting `10`.

> Otoh, if the txn shard's upper is not past `X` yet, then anything could happen and we have to wait until it does before saying anything about the snapshot

I think I'm probably confused as to how exactly it's supposed to work or asked the question incorrectly. If `X = 30` I thought that saying the last non-empty write was at `20` implied that the txn shard's upper was not past `X`.

Are the data shard uppers always equal to the last applied non-empty write on that shard? And is the txn-shard upper always greater than or equal to the last non-empty write of all shards (potentially greater than due to registrations and retractions)? 


Let me try and walk through the example in more detail:

1. Register shard `A` at ts `1`.
2. Register shard `B` at ts `2`.
3. Commit and apply write on shard `A` at ts `3`.
4. Commit and apply write on shard `B` at ts `4`.
5. Initiate snapshot of shard `A` at ts `10`.
  a. The timestamp does not get translated and stays at `10` because the txn shard upper is `5` which is not past `10` (I think).
  b. The snapshot is blocked because we have to wait until the txn shard upper advances past `10`.
6. Commit and apply write on shard `B` at ts `20`.
  a. Moving the txn shard upper to `21`.

My questions are: 

- Was I correct in saying the the timestamp does not get translated and stays at `10`? 
- If so, what mechanism informs the blocked snapshot that the logical upper has advanced and the snapshot can complete?

Reply:
(whoops initially typed that wrong, ignore the version of this comment in your email 😅)

Reply:
Ok, that clears everything up. I think what I was missing was that before we translate the timestamp we must first wait for the txn shard's upper to pass our timestamp (i.e. call `TxnsRead::update_gt(X)`).

Reply:
Cool! Called these three steps out explicitly in the doc